DMTN-229.tex

```diff
@@ -1,39 +1,435 @@
+It can handle raw files that store all the detector images in a single file or as one file per detector, and Butler does allow a single file artifact to be associated with multiple datasets in the Registry.
+In situations where the files may be ingested into Butler repositories multiple times, a facility is also provided to use ``sidecar'' JSON files containing extracted metadata, or even a JSON index file containing metadata for multiple files, since it much quicker to parse JSON than to extract and parse a FITS header, especially if that FITS file is in an object store.
+
+\subsection{Calibrations}
+
+As noted above, a \texttt{CALIBRATION} collection is a special type of collection that associates datasets with a validity range.
+When a calibration is requested the data coordinate of the dataset being calibrated is used and any time-based dimensions (such as \texttt{exposure}) specify which calibration should be chosen.
+
+Not all calibrations are calculated from datasets available to a Butler, for example, for LSSTCam the QE curves are calculated by the camera team and given to the pipelines team as part of the camera delivery.
+Additionally, some calibrations, such as defect masks, are relatively static, can be represented in a compact text format, and are useful to people outside of a Butler.
+We call such datasets ``curated calibrations'' and store them in Git repositories using text file formats and a directory layout that makes the validity ranges clear.
+It can be very convenient when setting up a new Butler for an instrument to be able to seed it with these curated calibrations without having to locate raw data files and reconstruct them from scratch, and infrastructure is provided to enable that, for example using the \texttt{butler write-curated-calibrations} command line tool.
+Generally, when these curated calibrations are stored the Butler converts them from text format to a binary format to make reading them more efficient; this is all handled by the formatter configured for the particular datastore.
+
+\section{The Pipeline System}
+
+\setcounter{footnote}{0}  % Latex only allows 9 footnotes with symbols.
+
+The fundamental unit of work for a pipeline is a \texttt{Task}.
+Each \texttt{Task} has a specialized configuration object attached to it (from the \texttt{pex\_config} package\footnote{\url{https://github.com/lsst/pex_config}}) and must provide a \texttt{run()} method that is the method that implements the algorithm.
+A \texttt{run()} method alone is not enough to integrate a task into a pipeline.
+To enable that it must be converted into a \texttt{PipelineTask}.
+
+A \texttt{PipelineTask} is special because it requires that the author declares the required input datasets and output datasets in a way that can be understood by a pipeline graph builder that is attached to a Butler registry.
+Critically, a \texttt{PipelineTask} must declare the dimensions that it is interested in.
```

Comment:
```suggestion
The connections also include \texttt{PipelineTask} must declare the dimensions over which the task runs.
The task's dimensions do not need to match those of its input or output datasets; for example, a task with dimensions \texttt{\{instrument, exposure\}} that takes an input with dimensions \texttt{\{instrument, exposure, detector\}} is a ``gather'' step that processes all detectors from that exposure together, but processes each exposure independently (and hence possibly in parallel).
```

Newcomers tend to be confused by what a task's dimensions represent, and I hope this example helps.